### ISSUE TITLE
Build error with gcc/libstdc++ 4.8.5

### ISSUE THREAD

**Issue by xavierba** [lightspark/lightspark#808]
I'm encountering the following error while building lightspark 0.8.5 for CentOS 7 which uses gcc/libstdc++ 4.8.5
```
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc_opcodes.cpp: In static member function 'static void lightspark::ABCVm::newClass(lightspark::call_context*, int)':
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc_opcodes.cpp:2851:79: error: no matching function for call to 'std::list<lightspark::Class_base*>::erase(std::_List_const_iterator<lightspark::Class_base*>&)'
    j = th->mi->context->root->applicationDomain->classesSuperNotFilled.erase(j);
                                                                               ^
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc_opcodes.cpp:2851:79: note: candidates are:
In file included from /usr/include/c++/4.8.2/list:64:0,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/swftypes.h:28,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/parsing/tags.h:26,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc.h:25,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc_opcodes.cpp:20:
/usr/include/c++/4.8.2/bits/list.tcc:108:5: note: std::list<_Tp, _Alloc>::iterator std::list<_Tp, _Alloc>::erase(std::list<_Tp, _Alloc>::iterator) [with _Tp = lightspark::Class_base*; _Alloc = std::allocator<lightspark::Class_base*>; std::list<_Tp, _Alloc>::iterator = std::_List_iterator<lightspark::Class_base*>]
     list<_Tp, _Alloc>::
     ^
/usr/include/c++/4.8.2/bits/list.tcc:108:5: note:   no known conversion for argument 1 from 'std::_List_const_iterator<lightspark::Class_base*>' to 'std::list<lightspark::Class_base*>::iterator {aka std::_List_iterator<lightspark::Class_base*>}'
In file included from /usr/include/c++/4.8.2/list:63:0,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/swftypes.h:28,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/parsing/tags.h:26,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc.h:25,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/abc_opcodes.cpp:20:
/usr/include/c++/4.8.2/bits/stl_list.h:1193:7: note: std::list<_Tp, _Alloc>::iterator std::list<_Tp, _Alloc>::erase(std::list<_Tp, _Alloc>::iterator, std::list<_Tp, _Alloc>::iterator) [with _Tp = lightspark::Class_base*; _Alloc = std::allocator<lightspark::Class_base*>; std::list<_Tp, _Alloc>::iterator = std::_List_iterator<lightspark::Class_base*>]
       erase(iterator __first, iterator __last)
       ^
/usr/include/c++/4.8.2/bits/stl_list.h:1193:7: note:   candidate expects 2 arguments, 1 provided
```
Builds fine on Fedora 33 (10.3.1) and Fedora 34/35/36 (11.2.1)
I know gcc 4.8.5 is quite old and I'm not expecting much, but maybe it can be fixed, or alternatively a minimal gcc version can be enforced.

**Comment by xavierba** [comment]
Actually the current minimal gcc version is 4.6.

**Comment by dbluelle** [comment]
Please test if the following patch fixes it:
```
--- src/scripting/abc_opcodes.cpp
+++ src/scripting/abc_opcodes.cpp
@@ -2836,20 +2836,20 @@ void ABCVm::newClass(call_context* th, int n)
 	assert_and_throw(asAtomHandler::isUndefined(ret2));
 	ASATOM_DECREF(ret2);
 	LOG_CALL(_("End of Class init ") << *mname <<" " <<ret);
 	RUNTIME_STACK_PUSH(th,asAtomHandler::fromObjectNoPrimitive(ret));
 	cinit->decRef();
 	if (ret->getGlobalScope()) // the variable on the Definition Object was set to null in class definition, but can be set to the real object now that the class init function was called
 		ret->getGlobalScope()->setVariableByQName(mname->name_s_id,mname->ns[0], ret,DECLARED_TRAIT);
 	th->mi->context->root->bindClass(className,ret);
 
-	auto j = th->mi->context->root->applicationDomain->classesSuperNotFilled.cbegin();
-	while (j != th->mi->context->root->applicationDomain->classesSuperNotFilled.cend())
+	auto j = th->mi->context->root->applicationDomain->classesSuperNotFilled.begin();
+	while (j != th->mi->context->root->applicationDomain->classesSuperNotFilled.end())
 	{
 		if((*j)->super == ret)
 		{
 			(*j)->copyBorrowedTraitsFromSuper();
 			j = th->mi->context->root->applicationDomain->classesSuperNotFilled.erase(j);
 		}
 		else
 			j++;
 	}
```

**Comment by xavierba** [comment]
Yes, the patch does help, but the build breaks a bit later:
 ```
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.cpp: In static member function 'static void lightspark::StringTools::_constructor(lightspark::asAtom&, lightspark::SystemState*, lightspark::asAtom&, lightspark::asAtom*, unsigned int)':
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.cpp:55:42: error: no matching function for call to 'std::locale::locale(std::string&)'
   th->currlocale = std::locale(localeName);
                                          ^
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.cpp:55:42: note: candidates are:
In file included from /usr/include/c++/4.8.2/bits/ios_base.h:41:0,
                 from /usr/include/c++/4.8.2/ios:42,
                 from /usr/include/c++/4.8.2/ostream:38,
                 from /usr/include/c++/4.8.2/iostream:39,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/compat.h:26,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/swftypes.h:23,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/asobject.h:23,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.h:23,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.cpp:20:
/usr/include/c++/4.8.2/bits/locale_classes.h:311:5: note: std::locale::locale(std::locale::_Impl*)
     locale(_Impl*) throw();
     ^
/usr/include/c++/4.8.2/bits/locale_classes.h:311:5: note:   no known conversion for argument 1 from 'std::string {aka std::basic_string<char>}' to 'std::locale::_Impl*'
/usr/include/c++/4.8.2/bits/locale_classes.h:178:7: note: template<class _Facet> std::locale::locale(const std::locale&, _Facet*)
       locale(const locale& __other, _Facet* __f);
       ^
/usr/include/c++/4.8.2/bits/locale_classes.h:178:7: note:   template argument deduction/substitution failed:
/builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.cpp:55:42: note:   cannot convert 'localeName' (type 'std::string {aka std::basic_string<char>}') to type 'const std::locale&'
   th->currlocale = std::locale(localeName);
                                          ^
In file included from /usr/include/c++/4.8.2/bits/ios_base.h:41:0,
                 from /usr/include/c++/4.8.2/ios:42,
                 from /usr/include/c++/4.8.2/ostream:38,
                 from /usr/include/c++/4.8.2/iostream:39,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/compat.h:26,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/swftypes.h:23,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/asobject.h:23,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.h:23,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/globalization/stringtools.cpp:20:
/usr/include/c++/4.8.2/bits/locale_classes.h:164:5: note: std::locale::locale(const std::locale&, const std::locale&, std::locale::category)
     locale(const locale& __base, const locale& __add, category __cat);
     ^
/usr/include/c++/4.8.2/bits/locale_classes.h:164:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/c++/4.8.2/bits/locale_classes.h:151:5: note: std::locale::locale(const std::locale&, const char*, std::locale::category)
     locale(const locale& __base, const char* __s, category __cat);
     ^
/usr/include/c++/4.8.2/bits/locale_classes.h:151:5: note:   candidate expects 3 arguments, 1 provided
/usr/include/c++/4.8.2/bits/locale_classes.h:137:5: note: std::locale::locale(const char*)
     locale(const char* __s);
     ^
/usr/include/c++/4.8.2/bits/locale_classes.h:137:5: note:   no known conversion for argument 1 from 'std::string {aka std::basic_string<char>}' to 'const char*'
/usr/include/c++/4.8.2/bits/locale_classes.h:126:5: note: std::locale::locale(const std::locale&)
     locale(const locale& __other) throw();
     ^
/usr/include/c++/4.8.2/bits/locale_classes.h:126:5: note:   no known conversion for argument 1 from 'std::string {aka std::basic_string<char>}' to 'const std::locale&'
/usr/include/c++/4.8.2/bits/locale_classes.h:117:5: note: std::locale::locale()
     locale() throw();
     ^
/usr/include/c++/4.8.2/bits/locale_classes.h:117:5: note:   candidate expects 0 arguments, 1 provided
```

**Comment by dbluelle** [comment]
OK, that should be fixed by replacing all (6) occurences of 
```
th->currlocale = std::locale(localeName);
```
with
```
th->currlocale = std::locale(localeName.c_str());
```
(these are all in files in /src/scripting/flash/globalization/ )

**Comment by xavierba** [comment]
Thanks, together with the previous patch, this is fixing the build.

There are a couple build warnings, pasting them here for information.
Only the first one is present in the gcc 10.2 build.

```
[ 13%] Building CXX object src/CMakeFiles/spark.dir/backends/security.cpp.o
cd /builddir/build/BUILD/lightspark-0.8.5/src && /usr/bin/c++  -DENABLE_CURL -DENABLE_LIBAVCODEC -DENABLE_RTMP -DGNASH_PATH=\"/usr/bin/gnash\" -DHAVE_ATOMIC -DHAVE_AVCODECID -DHAVE_AVCODEC_ALLOC_CONTEXT3 -DHAVE_AVCODEC_DECODE_AUDIO4 -DH
AVE_AVCODEC_DECODE_VIDEO2 -DHAVE_AVCODEC_OPEN2 -DHAVE_AVCODEC_RECEIVE_FRAME -DHAVE_AVCODEC_SEND_PACKET -DHAVE_AVFORMAT_CLOSE_INPUT -DHAVE_AVFORMAT_FIND_STREAM_INFO -DHAVE_AVIO_ALLOC_CONTEXT -DHAVE_AV_FRAME_ALLOC -DHAVE_AV_FRAME_GET_SAMP
LE_RATE -DHAVE_AV_FRAME_UNREF -DHAVE_AV_PACKET_UNREF -DHAVE_LIBSWRESAMPLE -DLIGHTSPARK_64 -DLS_DATADIR=\"/usr/share/lightspark\" -DPRIVATELIBDIR=\"/usr/lib64/lightspark\" -Dspark_EXPORTS -I/builddir/build/BUILD/lightspark-0.8.5 -I/usr/i
nclude/freetype2 -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/libpng15 -I/usr/include/uuid -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/us
r/include/fribidi -I/usr/include/SDL2 -I/usr/include/ffmpeg -I/builddir/build/BUILD/lightspark-0.8.5/src -I/builddir/build/BUILD/lightspark-0.8.5/src/scripting -I/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/jxrgluelib  -O2
 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -Wall -Wnon-virtual-dtor -Woverloaded-virtual -pipe -fvisibility=hidden -fvisibility-inl
ines-hidden -std=c++11 -Wdisabled-optimization -Wextra -Wno-unused-parameter -Wno-invalid-offsetof -g -O2 -DNDEBUG -fPIC   -o CMakeFiles/spark.dir/backends/security.cpp.o -c /builddir/build/BUILD/lightspark-0.8.5/src/backends/security.c
pp
/builddir/build/BUILD/lightspark-0.8.5/src/backends/locale.cpp: In member function 'void lightspark::LocaleManager::init()':
/builddir/build/BUILD/lightspark-0.8.5/src/backends/locale.cpp:32:6: note: variable tracking size limit exceeded with -fvar-tracking-assignments, retrying without
 void LocaleManager::init()
      ^
```
The one above is present too when building with gcc 11.2 (without the 2 patches added for the gcc 4.6 build).

```
[ 45%] Building CXX object src/CMakeFiles/spark.dir/scripting/flash/display3d/flashdisplay3dtextures.cpp.o
cd /builddir/build/BUILD/lightspark-0.8.5/src && /usr/bin/c++  -DENABLE_CURL -DENABLE_LIBAVCODEC -DENABLE_RTMP -DGNASH_PATH=\"/usr/bin/gnash\" -DHAVE_ATOMIC -DHAVE_AVCODECID -DHAVE_AVCODEC_ALLOC_CONTEXT3 -DHAVE_AVCODEC_DECODE_AUDIO4 -DH
AVE_AVCODEC_DECODE_VIDEO2 -DHAVE_AVCODEC_OPEN2 -DHAVE_AVCODEC_RECEIVE_FRAME -DHAVE_AVCODEC_SEND_PACKET -DHAVE_AVFORMAT_CLOSE_INPUT -DHAVE_AVFORMAT_FIND_STREAM_INFO -DHAVE_AVIO_ALLOC_CONTEXT -DHAVE_AV_FRAME_ALLOC -DHAVE_AV_FRAME_GET_SAMP
LE_RATE -DHAVE_AV_FRAME_UNREF -DHAVE_AV_PACKET_UNREF -DHAVE_LIBSWRESAMPLE -DLIGHTSPARK_64 -DLS_DATADIR=\"/usr/share/lightspark\" -DPRIVATELIBDIR=\"/usr/lib64/lightspark\" -Dspark_EXPORTS -I/builddir/build/BUILD/lightspark-0.8.5 -I/usr/i
nclude/freetype2 -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/libpng15 -I/usr/include/uuid -I/usr/include/libdrm -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/us
r/include/fribidi -I/usr/include/SDL2 -I/usr/include/ffmpeg -I/builddir/build/BUILD/lightspark-0.8.5/src -I/builddir/build/BUILD/lightspark-0.8.5/src/scripting -I/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/jxrgluelib  -O2
 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -Wall -Wnon-virtual-dtor -Woverloaded-virtual -pipe -fvisibility=hidden -fvisibility-inl
ines-hidden -std=c++11 -Wdisabled-optimization -Wextra -Wno-unused-parameter -Wno-invalid-offsetof -g -O2 -DNDEBUG -fPIC   -D__ANSI__ -DDISABLE_PERF_MEASUREMENT -Wno-endif-labels -I/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxr
lib/common/include/ -I/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/image/sys/ -o CMakeFiles/spark.dir/scripting/flash/display3d/flashdisplay3dtextures.cpp.o -c /builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/dis
play3d/flashdisplay3dtextures.cpp
In file included from /builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmspecstring.h:40:0,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/jxrgluelib/JXRMeta.h:35,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/jxrgluelib/JXRGlue.h:34,
                 from /builddir/build/BUILD/lightspark-0.8.5/src/scripting/flash/display3d/flashdisplay3dtextures.cpp:7:
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:334:48: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __readableTo(extent)    __declspec("SAL_readableTo("SPECSTRINGIZE(extent)")")
                                                ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:343:50: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __elem_readableTo(size)   __declspec("SAL_readableTo(elementCount("SPECSTRINGIZE(size)"))")
                                                  ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:350:50: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __byte_readableTo(size)   __declspec("SAL_readableTo(byteCount("SPECSTRINGIZE(size)"))")
                                                  ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:360:45: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __writableTo(size)   __declspec("SAL_writableTo("SPECSTRINGIZE(size)")")
                                             ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:367:50: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __elem_writableTo(size)   __declspec("SAL_writableTo(elementCount("SPECSTRINGIZE(size)"))")
                                                  ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:374:50: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __byte_writableTo(size)   __declspec("SAL_writableTo(byteCount("SPECSTRINGIZE(size)"))")
                                                  ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:457:61: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __inner_control_entrypoint(category) __declspec("SAL_entrypoint(controlEntry, "SPECSTRINGIZE(category)")")
                                                             ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:462:61: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __inner_data_entrypoint(category)    __declspec("SAL_entrypoint(dataEntry, "SPECSTRINGIZE(category)")")
                                                             ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:464:60: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __inner_success(expr)               __declspec("SAL_success("SPECSTRINGIZE(expr)")")
                                                            ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:466:60: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __inner_typefix(ctype)              __declspec("SAL_typefix("SPECSTRINGIZE(ctype)")")
                                                            ^
/builddir/build/BUILD/lightspark-0.8.5/src/3rdparty/jxrlib/common/include/wmsal.h:469:60: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wliteral-suffix]
     #define __inner_blocksOn(resource)          __declspec("SAL_blocksOn("SPECSTRINGIZE(resource)")")
```
The one above is _not_ present when building with gcc 11.2.
